docs/book/src/migration/guide.md

```diff
@@ -0,0 +1,192 @@
+# Migration Guide
+
+This document describes how to migrate a project from kubebuilder v1 to v2.
+
+We suggest you read [this doc](./v1vsv2.md) that compares the notable differences between v1 and v2.
+
+The recommended way of migration is to create a new v2 project and copy the business logic code.
+
+Please ensure you have followed the [installation guide](../quick-start.md#installation) to install the required
+components.
+
+We have an [example v1 project](https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/migration/testdata/gopath/project). It should look like the
+[v2 project in the cronjob tutorial](https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/cronjob-tutorial/testdata/project) after the migration.
+
+## Preparation
+
+First, we need to figure out what are the domain, group, version and kind for kubebuilder.
+
+In your kubebuilder v1 project, the `pkg` directory should look like this.
+
+```
+pkg/
+├── apis
+│   ├── addtoscheme_{your-group}_{your-version}.go
+│   ├── apis.go
+│   └── {your-group}
+│       ├── group.go
+│       └── {your-version}
+│           ├── {your-kind-in-lower-case}_types.go
+│           ├── {your-kind-in-lower-case}_types_test.go
+│           ├── doc.go
+│           ├── register.go
+│           ├── {your-version}_suite_test.go
+│           └── zz_generated.deepcopy.go
+├── controller
+└── webhook
+```
+
+In `register.go`, you can find something like
+```go
+SchemeGroupVersion = schema.GroupVersion{Group: {your-group}.{your-domain}, Version: {your-version}}
+```
+
+Now you have figured out what are your domain, group, version and kind.
+
+## Initialize the Project
+
+Run the following command
+
+```bash
+kubebuilder init --domain {your-domain}
+```
+
+## Migrate API and Controller
+
+Run the following command to scaffold the API and the controller. Please follow the prompt to scaffold the API and (or)
+the controller.
+
+```bash
+kubebuilder create api --group {your-group} --version {your-version} --kind {Your-kind}
+```
+
+You can run it multiple times if you have more than one API and controller. If you need multiple groups, please follow
+[this document](../TODO.md)
+
+
+### Migrate the API
+
+Now copy your API definition from `pkg/apis/{your-group}/{your-version}/{your-kind}_types.go` to
+`api/{your-version}/{your-kind}_types.go`. You only need to copy the implementation of your `Spec` and `Status`.
+
+You don't need to copy these markers from v1 project
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+```
+
+You need to ensure marker `// +kubebuilder:object:root=true` exists in v2 project and is the 2nd closest comment for
+your root object. e.g. `CronJob` and `CronJobList`
```

Comment:
be more concrete, otherwise this seems like an arcane incantation --

> we can replace the `+k8s:deepcopy-gen:interfaces` marker with `+kubebuilder:object:root=true`.

etc